src/pages/workspaces/workspace/useDeleteWorkspaceState.ts

```diff
@@ -35,83 +53,150 @@ export interface DeleteWorkspaceHookArgs {
+  )
 
   const deleteWorkspace = async () => {
+    if (isDeleteDisabledFromResources) {
+      throw new Error('Workspace contains non-deletable resources')
+    }
+
     try {
       setDeleting(true)
-      if (isGoogleWorkspace(hookArgs.workspace)) {
+      if (isGoogleWorkspace(hookArgs.workspace) && workspaceResources) {
         await Promise.all(
-          _.map(async app => await Ajax().Apps.app(app.cloudContext.cloudResource, app.appName).delete(), deletableApps)
+          _.map(async app => await Ajax(signal).Apps.app(app.cloudContext.cloudResource, app.appName).delete(), workspaceResources.deleteableApps)
         )
       }
-      await Ajax().Workspaces.workspace(workspaceInfo.namespace, workspaceInfo.name).delete()
+
+      await Ajax(signal).Workspaces.workspace(workspaceInfo.namespace, workspaceInfo.name).delete()
       hookArgs.onDismiss()
       hookArgs.onSuccess()
     } catch (error) {
-      reportError('Error deleting workspace', error)
       setDeleting(false)
+      await reportError('Error deleting workspace', error)
+    }
+  }
+
+  const checkAzureResources = async () => {
+    console.log('Checking azure resources...') // eslint-disable-line no-console
+    const appsInfo = await fetchWorkspaceResources(hookArgs.workspace)
+    setWorkspaceResources(appsInfo)
+
+    if (appsInfo.apps.length > 0 || appsInfo.runtimes.length > 0) {
+      console.log('Resources still present, rescheduling check...') // eslint-disable-line no-console
+      checkAzureResourcesTimeout.current = window.setTimeout(() => checkAzureResources(), WorkspaceResourceDeletionPollRate)
+    } else {
+      console.log('Resources gone.') // eslint-disable-line no-console
+      setDeletingResources(false)
+    }
+  }
+
+  const deleteWorkspaceResources = async () => {
+    if (isGoogleWorkspace(hookArgs.workspace)) {
+      throw new Error('Attempting to delete resources in an unsupported workspace')
+    }
+
+    if (workspaceResources && workspaceResources.nonDeleteableApps.length > 0) {
+      throw new Error('Workspace contains non-deletable apps')
+    }
+
+    if (workspaceResources && workspaceResources.nonDeleteableRuntimes.length > 0) {
+      throw new Error('Workspace contains non-deletable runtimes')
+    }
+
+    try {
+      setDeletingResources(true)
+      console.log(`Requesting app and runtime deletion for workspace ${workspaceInfo.workspaceId}`) // eslint-disable-line no-console
+      await Ajax(signal).Apps.deleteAllAppsV2(workspaceInfo.workspaceId)
+      await Ajax(signal).Runtimes.runtimeV2(workspaceInfo.workspaceId).deleteAll()
+      console.log('Resource deletions requested, starting poll.') // eslint-disable-line no-console
+
+      checkAzureResourcesTimeout.current = window.setTimeout(() => checkAzureResources(), WorkspaceResourceDeletionPollRate)
+    } catch (error) {
+      setDeletingResources(false)
+      throw error
```

Comment:
What then catches this error and displays something?

Reply:
I'll switch this to `reportError`; this is getting toward a question I had about the error handling on L164-L172 though -- should those be brought inside the `try` on L176 to ensure those get displayed too? If that's the case they probably shouldn't throw and should instead `reportError` + `return` as well I suppose.